src/Microsoft.AspNetCore.WebUtilities/KeyValueAccumulator.cs

```diff
@@ -9,24 +9,33 @@ namespace Microsoft.AspNetCore.WebUtilities
 {
     public struct KeyValueAccumulator
     {
-        private Dictionary<string, List<string>> _accumulator;
+        private Dictionary<string, StringValues> _accumulator;
 
         public void Append(string key, string value)
         {
             if (_accumulator == null)
             {
-                _accumulator = new Dictionary<string, List<string>>(StringComparer.OrdinalIgnoreCase);
+                _accumulator = new Dictionary<string, StringValues>(StringComparer.OrdinalIgnoreCase);
             }
-            List<string> values;
+
+            StringValues values;
             if (_accumulator.TryGetValue(key, out values))
             {
-                values.Add(value);
+                if (values.Count <= 1)
+                {
+                    _accumulator[key] = new string[] { values[0], value };
+                }
+                else
+                {
+                    var currentValues = values.ToArray();
+                    Array.Resize(ref currentValues, values.Count + 1);
```

Comment:
Will save allocs for values with counts  > 3 (or > 5 if you include lists internals); any idea on how often that occurs?
e.g. when I get into that situation I'll generally move to a json post as the array post is easier to reason about what's going on.